### ISSUE TITLE
Unable to push to a private GitHub repo from ci block

### ISSUE THREAD

**Issue by cjkent** [Shippable/support#4731]
I'm trying to push a commit to the checked-out repo in the `ci` block of my Shippable file. I'm seeing this error:

```
ERROR: Repository not found.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

Should I be able to push to the remote repo from the `ci` block?

The puzzling thing is that it succeeded once (I have the commit in the repo to prove it) but has never worked again even if everything is in the same apparent state.

Here is an example build:
https://app.shippable.com/github/OpenGamma/Utilities/runs/76/1/console

**Comment by Bharath92** [comment]
Could you please add the command `ssh -T git@github.com` before you try to push from the `ci` block? This is to check whether the authentication succeeds or not with the correct user.

**Comment by cjkent** [comment]
```Hi opengammacibot! You've successfully authenticated, but GitHub does not provide shell access.```

https://app.shippable.com/github/OpenGamma/Utilities/runs/78/1/console

**Comment by Bharath92** [comment]
It looks like the `opengammacibot` account is used for the `git push`. Could you please check whether `opengammacibot` has write access to the project which has to be `git push`ed?

If it doesn't, you could either:
- give write access to the project for the `opengammacibot` account **OR**
- delete the ssh key added from the `Subscriptions > Settings` page from the `opengammacibot` account and add it to an account which has write access to the project

http://docs.shippable.com/ci/ssh-keys/#ssh-keys

**Comment by cjkent** [comment]
AFAICT the `opengammacibot` account has no access to that repo at all. When I'm logged into GitHub as the bot I can't see the repo.

Which makes me wonder what credentials Shippable is using to clone the repo in the first place and why they're not the same as the ones available in the `ci` block.

**Comment by cjkent** [comment]
Should the git setup in the `ci` block be the same as the setup used to clone the repo?

**Comment by a-murphy** [comment]
A deploy key added to the project on GitHub is used to clone the repository for CI.  These keys do not have write access.

In order to push to a repository, you will need to set up write access from the job.  The easiest way to do this is to add the subscription SSH key (on the subscription settings page) to either the repository as a deploy key with write access or an account with access on GitHub.   You can also [add your own SSH key](http://docs.shippable.com/ci/ssh-keys/#adding-your-own-ssh-key) if you would prefer to use a different key.  Since the project deploy key has access to the repository, but not write permission, you may need to specify which key to use if you are adding your own deploy key and it continues to try to use the project deploy key.

**Comment by cjkent** [comment]
The only deploy key associated with the repo looks like it has read/write access. Or am I misunderstanding how it works?

![image](https://user-images.githubusercontent.com/98477/52401807-824ed980-2aba-11e9-9862-a1351b594f9e.png)

**Comment by rageshkrishna** [comment]
@cjkent Sorry about the confusion. Yes, the deploy keys should now have write access as you've discovered. They originally did not, but we changed that a while back. Let me try and explain exactly how we present keys for authentication.

When a command is executed in a build, we have `ssh-agent` loaded with two keys -- the subscription's key, and the project's deploy key. What happens when you use git is that ssh-agent presents these keys for authentication in exactly the same order as we added them. In your case, the subscription's key has been added to the `opengammacibot` GitHub user. As a result, authentication succeeds with the first key offered by ssh-agent. Unfortunately, this user fails authorization for this repository because it has no access to the repo on GitHub.

The _easiest_ way out at this point is to give `opengammacibot` permission to write to this repository.

There are other alternatives, but they require more work or may have some drawbacks. I'll list them here, but avoid the details for brevity. Let me know if you want to explore these options:

1. Use a completely new SSH/PEM integration, add it on GitHub and use this integration in your build. 

1. Don't use the subscription key with `opengammacibot`. However, if you're already using the key on `opengammacibot` for other reasons (submodules?), this isn't really an option.

I hope this clarifies how things are working. Happy to answer any more questions you might have.
